soc/arm/nxp_lpc/lpc55xxx/soc.c

```diff
@@ -58,59 +113,64 @@ const pll_setup_t pll0Setup = {
 
 static ALWAYS_INLINE void clock_init(void)
 {
-
-#if defined(CONFIG_SOC_LPC55S36)
-	/* Power Management Controller initialization */
+#if (CONFIG_SOC_LPC55S36)
+	/* Power Management Controller Initialization */
 	POWER_PowerInit();
 #endif
-
-#if defined(CONFIG_SOC_LPC55S06) || defined(CONFIG_SOC_LPC55S16) || \
-	defined(CONFIG_SOC_LPC55S28) || defined(CONFIG_SOC_LPC55S36) || \
-	defined(CONFIG_SOC_LPC55S69_CPU0)
     /*!< Set up the clock sources */
     /*!< Configure FRO192M */
 	/*!< Ensure FRO is on  */
 	POWER_DisablePD(kPDRUNCFG_PD_FRO192M);
 	/*!< Set up FRO to the 12 MHz, just for sure */
-	CLOCK_SetupFROClocking(12000000U);
+	CLOCK_SetupFROClocking(12000000);
 	/*!< Switch to FRO 12MHz first to ensure we can change the clock */
 	CLOCK_AttachClk(kFRO12M_to_MAIN_CLK);
 
-	/* Enable FRO HF(96MHz) output */
-	CLOCK_SetupFROClocking(96000000U);
+#if defined(CONFIG_INIT_PLL0) && !defined(CONFIG_SOC_LPC55S36)
+	/* Enable XTALHF clock */
+	POWER_DisablePD(kPDRUNCFG_PD_XTAL32M);
+	POWER_DisablePD(kPDRUNCFG_PD_LDOXO32M);
+#endif
 
-#ifdef CONFIG_INIT_PLL0
-	/*!< Ensure XTAL16M is on  */
-	PMC->PDRUNCFGCLR0 |= PMC_PDRUNCFG0_PDEN_XTAL32M_MASK;
-	PMC->PDRUNCFGCLR0 |= PMC_PDRUNCFG0_PDEN_LDOXO32M_MASK;
+#if !defined(CONFIG_TRUSTED_EXECUTION_NONSECURE)
+	/* Set Voltage for one of the fastest clock outputs: System clock output */
+	POWER_SetVoltageForFreq(DT_PROP(DT_CLOCKS_CTLR_BY_NAME(SYSCON_NODE, sysclk),
+					clock_frequency));
+	/* Set FLASH wait states for core */
+	CLOCK_SetFLASHAccessCyclesForFreq(DT_PROP(DT_CLOCKS_CTLR_BY_NAME(SYSCON_NODE,
+									 sysclk),
+								clock_frequency));
+#endif
 
-	/*!< Ensure CLK_IN is on  */
+	/*!< Ensure CLK_IN is on is either PLL is enabled  */
+#if ((DT_PROP(DT_CLOCKS_CTLR_BY_NAME(SYSCON_NODE, sysclk), clock_frequency) == 150000000) \
+	|| (defined(CONFIG_INIT_PLL0)))
+	CLOCK_SetupExtClocking(16000000);
```

Comment:
Rather than hardcoding the frequency and use of XTAL32M, could we make this oscillator another node in the devicetree.  Then DTS can control if it is enabled or not, and what the crystal frequency is (which is board-dependent).